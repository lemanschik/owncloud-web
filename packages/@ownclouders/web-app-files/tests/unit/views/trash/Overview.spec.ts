import TrashOverview from '../../../../src/views/trash/Overview.vue'
import {
  createStore,
  defaultComponentMocks,
  defaultPlugins,
  defaultStoreMockOptions,
  defaultStubs,
  mount,
  RouteLocation
} from 'web-test-helpers'
import { mock } from 'jest-mock-extended'
import { nextTick } from 'vue'

const spaceMocks = [
  {
    id: '1',
    name: 'admin',
    disabled: false,
    driveType: 'personal',
    getDriveAliasAndItem: () => '1'
  },
  {
    id: '2',
    name: 'Project space 1',
    disabled: false,
    driveType: 'project',
    getDriveAliasAndItem: () => '2'
  },
  {
    id: '3',
    name: 'Project space 2',
    disabled: false,
    driveType: 'project',
    getDriveAliasAndItem: () => '3'
  }
]

describe('TrashOverview', () => {
  it('should render no content message if no spaces exist', async () => {
    const { wrapper } = getWrapper({ spaces: [] })
    await wrapper.vm.loadResourcesTask.last
    expect(wrapper.find('no-content-message-stub').exists()).toBeTruthy()
  })
  it('should navigate to single space trash if only one space exists', () => {
    const { mocks } = getWrapper({ spaces: [spaceMocks[0]] })
    expect(mocks.$router.push).toHaveBeenCalledWith({
      name: 'files-trash-generic',
      params: { driveAliasAndItem: spaceMocks[0].getDriveAliasAndItem() },
      query: {}
    })
  })
  describe('view states', () => {
    it('shows the loading spinner during loading', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.find('oc-spinner-stub').exists()).toBeTruthy()
    })
    it('should render spaces list', async () => {
      const { wrapper } = getWrapper()
      await wrapper.vm.loadResourcesTask.last
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('sorting', () => {
    it('sorts by property name', async () => {
      const { wrapper } = getWrapper()
      await wrapper.vm.loadResourcesTask.last
      let sortedSpaces = []

      wrapper.vm.sortBy = 'name'
      await nextTick()
      sortedSpaces = wrapper.findComponent<any>({ name: 'oc-table' }).props().data
      expect(sortedSpaces.map((s) => s.id)).toEqual([
        spaceMocks[0].id,
        spaceMocks[1].id,
        spaceMocks[2].id
      ])

      wrapper.vm.sortDir = 'desc'
      await nextTick()
      sortedSpaces = wrapper.findComponent<any>({ name: 'oc-table' }).props().data
      expect(sortedSpaces.map((s) => s.id)).toEqual([
        spaceMocks[0].id,
        spaceMocks[2].id,
        spaceMocks[1].id
      ])
    })
    it('should set the sort parameters accordingly when calling "handleSort"', () => {
      const { wrapper } = getWrapper({ spaces: [spaceMocks[0]] })
      const sortBy = 'members'
      const sortDir = 'desc'
      wrapper.vm.handleSort({ sortBy, sortDir })
      expect(wrapper.vm.sortBy).toEqual(sortBy)
      expect(wrapper.vm.sortDir).toEqual(sortDir)
    })
  })
  describe('filtering', () => {
    it('shows only filtered spaces if filter applied', async () => {
      const { wrapper } = getWrapper()
      wrapper.vm.filterTerm = 'admin'
      await nextTick()
      expect(wrapper.vm.displaySpaces.length).toEqual(1)
      expect(wrapper.vm.displaySpaces[0].id).toEqual(spaceMocks[0].id)
    })
  })
})

function getWrapper({ spaces = spaceMocks } = {}) {
  const storeOptions = { ...defaultStoreMockOptions }
  const store = createStore(storeOptions)
  storeOptions.modules.runtime.modules.spaces.getters.spaces.mockReturnValue(spaces)
  const mocks = {
    ...defaultComponentMocks({
      currentRoute: mock<RouteLocation>({ name: 'trash-overview' })
    })
  }

  return {
    mocks,
    wrapper: mount(TrashOverview, {
      global: {
        stubs: { ...defaultStubs, NoContentMessage: true },
        mocks,
        plugins: [...defaultPlugins(), store]
      }
    })
  }
}
