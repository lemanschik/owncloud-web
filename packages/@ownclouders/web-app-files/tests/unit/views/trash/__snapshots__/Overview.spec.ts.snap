// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrashOverview view states should render spaces list 1`] = `
<div class="oc-flex">
  <div class="files-view-wrapper oc-width-expand">
    <div id="files-view">
      <app-bar-stub breadcrumbs="[object Object]" breadcrumbscontextactionsitems="" hasbulkactions="false" hasfileextensions="false" hashiddenfiles="false" haspagination="false" hassharesnavigation="false" hassidebartoggle="false" hasviewoptions="false" showactionsonselection="false" sidebaropen="false" viewmodes=""></app-bar-stub>
      <div class="oc-ml-m oc-my-s">
        <label class="oc-label" for="spaces-filter">Search</label>
        <div class="oc-position-relative">
          <!--v-if-->
          <input aria-invalid="false" autocomplete="off" class="oc-text-input oc-input oc-rounded" id="spaces-filter" type="text">
          <!--v-if-->
        </div>
        <!--v-if-->
      </div>
      <table class="oc-table oc-table-hover oc-table-sticky spaces-table">
        <thead class="oc-thead">
          <tr class="oc-table-header-row" tabindex="-1">
            <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-icon oc-pl-s" style="top: 0px;">
              <span class="oc-table-thead-content header-text"></span>
              <!--v-if-->
            </th>
            <th aria-sort="ascending" class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-name oc-pr-s" style="top: 0px;">
              <span class="oc-table-thead-content header-text">Name</span>
              <button aria-label="Sort by name" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-button-sort" type="button">
                <!-- @slot Content of the button -->
                <span class="oc-icon oc-icon-s oc-icon-passive">
                  <!---->
                </span>
              </button>
            </th>
          </tr>
        </thead>
        <tbody class="has-item-context-menu">
          <tr class="oc-tbody-tr oc-tbody-tr-1" data-item-id="1" draggable="false" tabindex="-1">
            <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-td oc-table-data-cell oc-table-data-cell-icon oc-pl-s">
              <span class="oc-icon oc-icon-m oc-icon-passive oc-pl-m">
                <!---->
              </span>
            </td>
            <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-name oc-pr-s">
              <router-link-stub class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-display-block" to="[object Object]">Personal</router-link-stub>
            </td>
          </tr>
          <tr class="oc-tbody-tr oc-tbody-tr-2" data-item-id="2" draggable="false" tabindex="-1">
            <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-td oc-table-data-cell oc-table-data-cell-icon oc-pl-s">
              <span class="oc-icon oc-icon-m oc-icon-passive oc-pl-m">
                <!---->
              </span>
            </td>
            <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-name oc-pr-s">
              <router-link-stub class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-display-block" to="[object Object]">Project space 1</router-link-stub>
            </td>
          </tr>
          <tr class="oc-tbody-tr oc-tbody-tr-3" data-item-id="3" draggable="false" tabindex="-1">
            <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-td oc-table-data-cell oc-table-data-cell-icon oc-pl-s">
              <span class="oc-icon oc-icon-m oc-icon-passive oc-pl-m">
                <!---->
              </span>
            </td>
            <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-name oc-pr-s">
              <router-link-stub class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-display-block" to="[object Object]">Project space 2</router-link-stub>
            </td>
          </tr>
        </tbody>
        <tfoot class="oc-table-footer">
          <tr class="oc-table-footer-row">
            <td class="oc-table-footer-cell" colspan="2">
              <!-- @slot Footer of the table -->
              <div class="oc-text-nowrap oc-text-center oc-width-1-1 oc-my-s">
                <p class="oc-text-muted">3 trashes in total</p>
                <!--v-if-->
              </div>
            </td>
          </tr>
        </tfoot>
        <!--v-if-->
      </table>
    </div>
  </div>
</div>
`;
